update/main.go

```diff
@@ -0,0 +1,133 @@
+package update
+
+import (
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+	"trellis-cli/github"
+
+	"github.com/hashicorp/go-version"
+	"github.com/mattn/go-isatty"
+	"gopkg.in/yaml.v2"
+)
+
+type Notifier struct {
+	CacheDir string
+	Repo     string
+	Version  string
+	Client   *http.Client
+}
+
+type StateEntry struct {
+	CheckedForUpdateAt time.Time       `yaml:"checked_for_update_at"`
+	LatestRelease      *github.Release `yaml:"latest_release"`
+}
+
+func (n *Notifier) CheckForUpdate() (*github.Release, error) {
+	if !n.shouldCheckForUpdate() {
+		return nil, nil
+	}
+
+	stateFilePath := filepath.Join(n.CacheDir, "state.yml")
+	latestRelease, err := n.getLatestReleaseInfo(stateFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if versionGreaterThan(latestRelease.Version, n.Version) {
+		return latestRelease, nil
+	}
+
+	return nil, nil
+}
+
+func (n *Notifier) shouldCheckForUpdate() bool {
+	if os.Getenv("TRELLIS_FORCE_UPDATE_NOTIFIER") != "" {
+		return true
+	}
+
+	if n.Repo == "" || n.CacheDir == "" {
+		return false
+	}
+
+	// skip completion commands
+	if os.Getenv("COMP_LINE") != "" {
+		return false
+	}
+
+	if os.Getenv("TRELLIS_NO_UPDATE_NOTIFIER") != "" {
+		return false
+	}
+
+	// skip on common CI environments
+	if os.Getenv("CI") != "" {
+		return false
+	}
+
+	// skip on non-terminals
+	fd := os.Stdout.Fd()
+	if !(isatty.IsTerminal(fd) || isatty.IsCygwinTerminal(fd)) {
+		return false
+	}
+
+	return true
+}
+
+func (n *Notifier) getLatestReleaseInfo(stateFilePath string) (*github.Release, error) {
+	stateEntry, err := getStateEntry(stateFilePath)
+	if err == nil && time.Since(stateEntry.CheckedForUpdateAt).Hours() < 24 {
+		return stateEntry.LatestRelease, nil
+	}
```

Comment:
Should we force to fetch from GitHub if `TRELLIS_FORCE_UPDATE_NOTIFIER` is set?

Reply:
🤔 it's a decent idea but it breaks my usage of that flag unfortunately. It mainly exists for testing purposes to get past the "is TTY?" check.

Reply:
Maybe add a `force` param on `CheckForUpdate`?

Reply:
Your idea was better than the env flag though as a direct replacement. So I refactored `TRELLIS_FORCE_UPDATE_NOTIFIER` to instead be a `ForceCheck` attribute on the notifier struct. And when I was redoing the tests, it exposed an issue 🎉  All fixed up now